homeassistant/components/dhcp/__init__.py

```diff
@@ -55,13 +56,22 @@
 _LOGGER = logging.getLogger(__name__)
 
 
-class DhcpServiceInfo(TypedDict):
+@dataclass
+class DhcpServiceInfo:
     """Prepared info from dhcp entries."""
 
     ip: str
     hostname: str
     macaddress: str
 
+    def __getitem__(self, name: str) -> Any:
+        """
+        Allow property access by name for compatibility reason.
+
+        Deprecated, and will be removed in version 2022.6.
+        """
+        return getattr(self, name)
```

Comment:
If it is deprecated we should make sure it throws a warning.

Reply:
@bdrace said "I think we want to wait to add the warning until after everything in core is updated in case we aren't able to complete transitioning them all before the next release"

Are you saying that we should not consider this a breaking change, and only mark is as deprecated/breaking change when we put the warning in? (most likely after 2021.12 is out)

Reply:
> I think we want to wait to add the warning until after everything in core is updated in case we aren't able to complete transitioning them all before the next release

In that case, we should document it as deprecated IMHO.

Reply:
Should we add a `debug` log entry, and upgrade it to `warning` after `2021.12` is out?
Then:
- we have it documented via the release notes (ie. breaking change).
- we have it documented in the python description.
- we don't have any commented code.

Maybe we could also add a TODO to remind us to upgrade to a warning (but I think the TODO is not recommended in HA).

Reply:
`debug` and upgrade to `warning` after we have everything in core converted seems like a good transition plan to me.

Reply:
We should consider it a breaking change right away since we want to make sure we maximize the time period where dev are aware and conversions to the dataclass accessors can happen

Reply:
@frenck do you agree with @bdraco before I retrigger the Actions?

Reply:
Yes, starting it's as breaking change and using debug level sounds sane for starters.
PS: We should make sure it only fires the log message once for each instance (to prevent floods).